File: fe/src/types/Assignment.ts
import { Dayjs } from 'dayjs'
import { BaseUser } from './User'

export interface Assignment {
  id: number
  title: string
  author: BaseUser
  startDate: Dayjs
  deadline: Dayjs
  submissions: Submission[]
  studentIds: number[]
  description: string
  maximumGrade: number
}

export interface AssignmentResponseDto {
  id: number
  title: string
  author: BaseUser
  startDate: string
  deadline: string
  students: BaseUser[]
  submissions: Submission[]
  description: string
  maximumGrade: number
}

export interface AssignmentRequestDto {
  title: string
  authorId: number
  startDate: string
  deadline: string
  studentIds: number[]
  submissions: Submission[]
  description: string
  maximumGrade: number
}

export interface Submission {
  studentId: number
  homeworkURL: string
  grade: number
  feedback: string
}

File: mp/src/main/java/ro/ubb/mp/dao/model/Assignment.java
package ro.ubb.mp.dao.model;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import javax.persistence.*;
import java.sql.Timestamp;
import java.util.List;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Entity(name = "assignments")
public class Assignment {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToMany(cascade = {
            CascadeType.PERSIST,
            CascadeType.MERGE
    })
    @JoinTable(name = "assignment_student_xref",
            joinColumns = @JoinColumn(name = "assignment_id"),
            inverseJoinColumns = @JoinColumn(name = "user_id")
    )
    private List<User> students;

    @ManyToOne(fetch = FetchType.LAZY)
    private User author;

    private String title;
    private Timestamp startDate;
    private Timestamp deadline;
    private String description;
    private Double maximumGrade;

    @ManyToMany(cascade = {
            CascadeType.PERSIST,
            CascadeType.MERGE
    })
    @JoinTable(name = "submission_student_xref",
            joinColumns = @JoinColumn(name = "assignment_id"),
            inverseJoinColumns = @JoinColumn(name = "submission_id")
    )
    private List<Submission> submissions;
}
